### ISSUE TITLE
Unicode emojis print wrong

### ISSUE THREAD

**Issue by jankatins** [wez/wezterm#1607]
### What Operating System(s) are you seeing this problem on?

Linux Wayland

### WezTerm version

20220130-201827-a8aafaeb

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

I tried the examples in https://darrenburns.net/posts/emoji-in-the-terminal/ to see how wezterm renders the emojis and it doesn't look nice:

```py
>>> print("\U0001F468\u200D\U0001F467\u200D\U0001F466")
👨‍👧‍👦
>>> print("👨\n12")
👨
12
>>> print("\U0001F4A3hello")
💣hello
>>> print("\U0001F6E5hello")
🛥hello
>>> print("\U0001F3F5\uFE0Fhello")
🏵️hello
```

On my system, the above looks like this:

![image](https://user-images.githubusercontent.com/890156/151789624-7b42c9ae-2da3-4e93-a668-f86757af0a64.png)

* It seems the emojis are printed with one char wide, some with only parts (faces), some scaled down (ship?)
* Some are still considered two chars wide as the hello after the bomb starts two chars away?

It seems this sentence (from theblog post) applies:

> All terminal emulators I tested consider codepoints with an "East Asian Width" of N to have cell width of 1. This is incorrect in the case of Emoji Presentation Sequences - Unicode recommends they should be always treated as "East Asian Wide" (W).

### To Reproduce

Run the above script

### Configuration

--

### Expected Behavior

The icons are printed in a nice way.

**Comment by wez** [comment]
This stuff is influenced by https://wezfurlong.org/wezterm/config/lua/config/unicode_version.html but may be a bit wonky at the moment as a result of https://github.com/wez/wezterm/commit/2deba1ece49a52a26188f70a86b375695281962a

**Comment by wez** [comment]
bisected this regression to 0da3e159a4f860c67422f7a4a3d39b2d3b59f82b

**Comment by jankatins** [comment]
Do I understand 5f2c905db86353040c63ee9335cf69ca9fd5e99c correctly that I should switch to unicode version 14? Is there some guidelines when which version is better? 

> For example, there were a number of unicode code points that had their width changed between Unicode version 8 and version 9. This wouldn't be an issue if all software was simultaneously aware of the change, but the reality is that there is a lot of older software out there, and that even if your local system is fully up to date, you might connect to a remote system vis SSH that is running applications that use a different version of unicode than your local system.

E.g. I do not understand in which situation I would prefer version 9?  It seems that version came out 2016, so I would have guessed that even connecting to LTS linuxes should have recent fonts there? (also I would guess that most wezterm users are version update junkies :-))

**Comment by jankatins** [comment]
There seems tobe still something wrong here: the ship is still too small in both unicode versions (tried with 20220201-120751-e409968f)

![image](https://user-images.githubusercontent.com/890156/152040944-8196ea89-a77f-4146-aa31-50da2abf74cc.png)

<details>
  <summary>Script (adjusted from whats in the repo):</summary>


```bash
#!/bin/zsh

# Save initial unicode version
local label=$(uuidgen)
printf "\e]1337;UnicodeVersion=push %s\e\\" $label

for version in 9 14 ; do
  echo "\n### Unicode Version $version\n"
  printf "\e]1337;UnicodeVersion=$version\e\\"

  echo -e "\u270c|  Victory hand, text presentation by default"
  echo -e "\u270c\ufe0e|  Victory hand, explicit text presentation"
  echo -e "\u270c\ufe0f|  Victory hand, explicit emoji presentation"

  echo -e "\u270a|  Raised fist, emoji presentation by default"
  echo -e "\u270a\ufe0e|  Raised fist, explicit text presentation (invalid; no effect)"
  echo -e "\u270a\ufe0f|  Raised fist, explicit emoji presentation"

  echo -e "\u2716|  Multiply, text presentation by default"
  echo -e "\u2716\ufe0e|  Multiply, explicit text presentation"
  echo -e "\u2716\ufe0f|  Multiply, explicit emoji presentation"

  echo -e "\U0001F468\u200D\U0001F467\u200D\U0001F466|"
  echo -e "\U0001F6E5|"
  echo "👨|"
  echo "12|"
done

# Restore saved unicode version
printf "\e]1337;UnicodeVersion=pop %s\e\\" $label
```
</details>

**Comment by wez** [comment]
Yeah, I put some math in the wrong loop.  Should be fixed now in main!

**Comment by abzcoding** [comment]
it seems to be wrong for me as well

<img width="1589" alt="Screen Shot 2022-02-02 at 10 56 14 AM" src="https://user-images.githubusercontent.com/10992695/152111393-f1ff0e53-c590-4e0c-be6b-852d46ee21f4.png">

i'm on the latest nightly version
<img width="269" alt="Screen Shot 2022-02-02 at 10 56 33 AM" src="https://user-images.githubusercontent.com/10992695/152111417-1aa26faf-8102-4128-9b82-61e832c4f8ac.png">

here is the `wt-record` output
https://gist.github.com/abzcoding/abfb39bfea6989f83e546f1fa1ae325a

**Comment by jankatins** [comment]
The ship is still one char instead of two (screenshot from a mac, but current version: 20220201-222722-98d2400f):

![image](https://user-images.githubusercontent.com/890156/152137264-ef64e6b7-3e22-446a-98d3-c92aae97b178.png)

**Comment by wez** [comment]
> The ship is still one char instead of two

It has a column width of 1, so it is correct that it occupies a single column.

With the default value for https://wezfurlong.org/wezterm/config/lua/config/allow_square_glyphs_to_overflow_width.html it will overflow into two columns when it is followed by a space.

> should switch to unicode version 14? Is there some guidelines when which version is better?

The question is really whether the programs that you use all agree on the unicode version.
If they don't then you'll see unexpected widths or other rendering glitches/artifacts.

When I initially added support for emoji variation selectors I made that the default behavior,
but I had a series of bug reports about incompatible software, so the sane way forward from
a support perspective was to set the default to be more compatible until more things have
caught up.

If setting the unicode version to 14 works fine for the programs that you use, then go for it!

**Comment by wez** [comment]
@abzcoding I can't replay that typescript on either linux or macos :-/ Can you reduce the issue to something like a simple printf or echo statement?

**Comment by wez** [comment]
@abzcoding or maybe use a tool like https://asciinema.org/ to capture what you're seeing?

**Comment by abzcoding** [comment]
> @abzcoding I can't replay that typescript on either linux or macos :-/ Can you reduce the issue to something like a simple printf or echo statement?

it's from mac os, yes i'll try to use [ascii](https://asciinema.org) or use a simple echo to help you out,
basically this picture, the left pane is https://github.com/wez/wezterm/commit/caf504ee819ac395aec8a7d9af131b75af1f9318 and the right pane is https://github.com/wez/wezterm/commit/98d2400fbaddacc725e607d47fb49e9bd39bc940
<img width="1914" alt="Screen Shot 2022-02-02 at 5 41 31 PM" src="https://user-images.githubusercontent.com/10992695/152170097-0c5cff52-5f3a-4f5f-965b-5a226083c649.png">

**Comment by abzcoding** [comment]
@wez 🤞 I think it should be fixed now, tyvm for your work on this, but let's wait for OP as well
<img width="532" alt="Screen Shot 2022-02-02 at 11 43 16 PM" src="https://user-images.githubusercontent.com/10992695/152229973-41b4e1d0-9d5e-401c-bc2f-9c8609105394.png">

**Comment by jankatins** [comment]
Re ship only one char wide: the linked blog post complained that it is displayed (as configured) as 1 char, but according to the blog, this should be ignored and displayed as 2 chars wide.

> All terminal emulators I tested consider codepoints with an "East Asian Width" of N to have cell width of 1. This is incorrect in the case of Emoji Presentation Sequences - Unicode recommends they should be always treated as "East Asian Wide" (W).

The recommendation links to http://www.unicode.org/reports/tr11/tr11-38.html#Recommendations
> [UTS51] emoji presentation sequences behave as though they were East Asian Wide, regardless of their assigned East_Asian_Width property value.

**Comment by wez** [comment]
If you put `\ufe0f` on the end of the sequence, that is the explicit emoji presentation variation selector and it should cause them to be 2 cells wide when wezterm is running with an appropriate unicode version selected.  Otherwise they default to the default presentation width (as returned by wcwidth) for the sequence.

**Comment by jankatins** [comment]
Now there is something broken with the ship emoji (mac, 20220202-084610-02426167, same mac as comment 12h ago):

![image](https://user-images.githubusercontent.com/890156/152245321-7e3a5a2c-79f2-4e07-bde4-9ec86285b386.png)

the two funny chars are
```
  echo -e "\U0001F6E5|"
  echo -e "\U0001F6E1|"
```

(the second emoji is the shield emoji, which https://github.com/ridiculousfish/widecharwidth/issues/11 is about)

**Comment by jankatins** [comment]
And it's fixed again: 
![image](https://user-images.githubusercontent.com/890156/152308654-aba858af-e866-4f60-abb4-4784986beb22.png)

Given the explanation above and in https://github.com/ridiculousfish/widecharwidth/issues/11#issuecomment-1028654433 (both the shield and the motorboat should display as one char in the terminal as they default to "text representation" and not "emoji presentation"), I guess this issue can be closed as fixed?

(also a lot of thanks for correcting me here and explaining why I got it wrong! Thanks!)

**Comment by wez** [comment]
> Given the explanation above and in [ridiculousfish/widecharwidth#11 (comment)](https://github.com/ridiculousfish/widecharwidth/issues/11#issuecomment-1028654433) (both the shield and the motorboat should display as one char in the terminal as they default to "text representation" and not "emoji presentation"), I guess this issue can be closed as fixed?

Yep!
 
> (also a lot of thanks for correcting me here and explaining why I got it wrong! Thanks!)

You're welcome!  This stuff is a PITA and often hard to follow!

> (also really strange: the wezterm version is the same, but the binary got updated (indicating the file brew downloads was changed?) and it does display differently. No idea if this was something else or the binary was updated?)

Weird!

**Comment by faho** [comment]
To add some more context here coming from the widecharwidth side, after I've read up some more on it:

Typically, the width a codepoint takes up is the width that EastAsianWidth assigns it. Typically this is "narrow" (1 cell) or "wide" (2 cells) with some special cases for ambiguous codepoints.

The bit of Unicode that "governs" this is [UAX#11: East Asian Width](https://unicode.org/reports/tr11/).

Now, since Unicode version 9 there's an exception there that relates to emoji. The current wording is:

> [[UTS51](https://www.unicode.org/reports/tr41/tr41-28.html#UTS51)] emoji presentation sequences behave as though they were East Asian Wide, regardless of their assigned East_Asian_Width property value.

You'll see this links to UTS51 which explains what an "emoji presentation sequence" is, and that is:

> emoji presentation sequence — A variation sequence consisting of an emoji character followed by a emoji presentation selector.

Where the "emoji presentation selector" is VS16. So this would be U+1F6E5 U+FE0F, but *not* U+1F6E5 by itself.

Technically this does *not* apply to _any_ emoji by itself, but it seems in practice most things consider emoji that have the "Emoji_Presentation" property (which essentially makes VS16 implied) for this as well.

So that is the interpretation we picked: An emoji is wide if it has Emoji_Presentation or is followed by VS16 (well, technically the wcwidth API can't express it but you can typically hack around it by assigning it a width of 1. I'm working on a better solution). If it isn't, we go by EastAsianWidth.

Now, we could also assume any "Emoji" is automatically wide, possibly unless followed by VS15 (the counterpoint to VS16 that selects "text presentation"). The issue with that?

Here are some emoji:

- `#`, `*`
- `5` (and 0-9)
- `🌶` (U+1F336 Hot Pepper, which is typically rendered as narrow in terminals)

There are apparently 1179 codepoints that are marked as Emoji but not Emoji_Presentation and have an EastAsianWidth other than 2.

So, I realize this is imperfect, but this is simply the best interpretation that we can come up with at the moment with an unclear standard.

**Comment by faho** [comment]
Okay, careful re-reading of UAX#11 turns up:

>[ED4](https://unicode.org/reports/tr11/#ED4). East Asian Wide (W): All other characters that are always wide. These characters occur only in the context of East Asian typography where they are wide characters (such as the Unified Han Ideographs or Squared Katakana Symbols). This category includes characters that have explicit halfwidth counterparts, along with characters that have the [[UTS51](https://www.unicode.org/reports/tr41/tr41-28.html#UTS51)] property Emoji_Presentation, with the exception of characters that have the [[UCD](https://www.unicode.org/reports/tr41/tr41-28.html#UCD)] property Regional_Indicator

So our interpretation is correct.
